test/tests.bats

```diff
@@ -53,8 +53,12 @@ setup_file() {
   docker cp "${PRIVATE_CONFIG}/sieve/dovecot.sieve" mail:/var/mail/localhost.localdomain/user1/.dovecot.sieve
 
   # this relies on the checksum file being updated after all changes have been applied
+  wait_for_changes_to_be_detected_in_container mail
```

Comment:
This seems to be the probable cause of the [current test failure](https://github.com/docker-mailserver/docker-mailserver/actions/runs/3169501301/jobs/5161461988#step:5:225)?:

```
not ok 217 setup_file failed
# (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 43,
#  from function `wait_until_change_detection_event_completes' in file test/test_helper/common.bash, line 207,
#  from function `setup_file' in test file test/tests.bats, line 57)
#   `wait_until_change_detection_event_completes mail' failed
# 93e41b2dace3e1093365580c698cef79767212ec5d1abae8aac795be5e32fd5a
# [   INF   ]  mail.my-domain.com is up and running
# Timed out on command: __is_changedetector_processing
# mail
# bats warning: Executed 217 instead of expected 318 tests
make: *** [Makefile:46: tests] Error 1
```

Looking at the line below it which was my new replacement for this method you added, we can see that this doesn't seem necessary?:

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/test/test_helper/common.bash#L204-L215

My method should notice if there is an existing change detection in progress, and if there is wait for it to complete for 60 seconds before timing out. If there isn't one, then it provides an additional 60 second window to wait for a change event to begin. Both should be more than enough time.

The old method that you added here, instead monitors for a file change to happen, but doesn't hook into the change detection service itself:

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/test/test_helper/common.bash#L168-L174

That has me a bit confused though. I can see from the error that it failed on this check:

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/test/test_helper/common.bash#L206-L207

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/test/test_helper/common.bash#L187

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/test/test_helper/common.bash#L190-L198

So a change was detected, but 60 seconds we got no feedback from the change detection service? I guess it was not running perhaps?...

It doesn't seem like your addition here is overlapping or causing any problems itself, `cmp --silent -- <(_monitored_files_checksums) "${CHKSUM_FILE}"` doesn't cause any conflict with the change detection service. Only making it apparent that a change did occur and is detectable, just that there is something wrong with checking change-detector logs during some test runs.

---

I'm curious if ClamAV has anything to do with that :thinking: I know it can be a RAM hog at times, and that may make the tests less reliable the longer it's been since we built an image with a fresh ClamAV update due to caching.

I have noticed in local testing before that ClamAV used over 6GB of RAM once (_I cannot recall if that was compressed with zram or the uncompressed amount_), I do know that after building a new image again without ClamAV cached it resolved the problem and I could run tests again.

If this is the problem, the Github Action CI machines have around 4-8GB RAM available IIRC, and it's possible that the system OOM-killer is also kicking in. So if the change detection service did start, a process needed (possibly the service itself or supervisord or docker was killed/restarted. I know killing ClamAV processes that are using too much RAM would just result in restarting that process and the memory usage quickly escalates again.

It also taxes the CPU slowing everything down I think (which can also happen with kernel kswapd if there is lots of memory pressure going on, even without swap, which historically is known to delay the OOM-killer, we are still using older Ubuntu CI runners.

---

I am not sure how many tests in `tests.bats` need ClamAV coverage or are specific to it. A solution might be to move those to a separate test file?

Alternatively, if we shift this to near the end of the Dockerfile, we could have our tests invalidate the ClamAV cache and pull a fresh image. I think BuildKit allows for invalidating cache for specific build stages:

https://github.com/docker-mailserver/docker-mailserver/blob/4ab23061a173c65d4d969881d0ae240abc5b4bf9/Dockerfile#L42-L44

Reply:
I have started outsourcing ClamAV-related tests. We'll see how it goes.